src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -454,156 +475,73 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
-            if (newName != oldCmd.Name())
+            const auto cmdImpl{ get_self<implementation::Command>(cmd) };
+            if (cmd.Origin() == OriginTag::User && cmd.ID().empty())
             {
-                // The new name differs from the old name,
-                // update our name.
-                auto oldCmdImpl{ get_self<Command>(oldCmd) };
-                oldCmdImpl->Name(newName);
+                // the user did not define an ID for their non-nested, non-iterable, valid command - generate one for them
+                cmdImpl->GenerateID();
             }
-        }
 
-        // Update maskingCmd:
-        //   We have a Command that is masking one from a parent layer.
-        //   We need to ensure that this has the correct name. That way,
-        //    we can return an accumulated view of a Command at this layer.
-        //   This differs from oldCmd which is mainly used for serialization
-        //    by recording the delta of the Command in this layer.
-        if (maskingCmd)
-        {
-            // This command has a name, check if it's new.
-            if (newName != maskingCmd.Name())
+            // only add to the _ActionMap if there is an ID
+            if (auto cmdID = cmd.ID(); !cmdID.empty() && cmd.ActionAndArgs().Action() != ShortcutAction::Invalid)
             {
-                // The new name differs from the old name,
-                // update our name.
-                auto maskingCmdImpl{ get_self<Command>(maskingCmd) };
-                maskingCmdImpl->Name(newName);
+                // in the legacy scenario, a user might have several of the same action but only one of them has defined an icon or a name
+                // eg. { "command": "paste", "name": "myPaste", "keys":"ctrl+a" }
+                //     { "command": "paste", "keys": "ctrl+b" }
+                // once they port over to the new implementation, we will reduce it to just one Command object with a generated ID
+                // but several key binding entries, like so
+                //     { "command": "newTab", "id": "User.paste" } -> in the actions map
+                //     { "keys": "ctrl+a", "id": "User.paste" }    -> in the keybindings map
+                //     { "keys": "ctrl+b", "id": "User.paste" }    -> in the keybindings map
+                // however, we have to make sure that we preserve the icon/name that might have been there in one of the command objects
+                // to do that, we check if this command we're adding had an ID that was generated
+                // if so, we check if there already exists a command with that generated ID, and if there is we port over any name/icon there might be
+                // (this may cause us to overwrite in scenarios where the user has an existing command that has the same generated ID but
+                //  performs a different action or has different args, but that falls under "play stupid games")
+                if (cmdImpl->IdWasGenerated())

```

Comment:
when do we track fixups?

Reply:
I've tried to consolidate tracking fixups to `ActionMap::LayerJson` in `ActionMapSerialization.cpp`, so only that function sets that flag and handles the logic for when to set it. Everything else shouldn't need to set it.